pkg/controllers/observability-ui/observability-ui-plugin/controller.go

```diff
@@ -0,0 +1,218 @@
+type Options struct {
+	PluginsConf ObservabilityUIPluginsConfiguration
+}
+
+// RBAC for managing ObservabilityUIPlugins
+// +kubebuilder:rbac:groups=observabilityui.rhobs,resources=observabilityuiplugins,verbs=get;list;watch;create;update;delete;patch
+// +kubebuilder:rbac:groups=observabilityui.rhobs,resources=observabilityuiplugins/status,verbs=get;update
+
+// RBAC for managing observability ui plugin objects
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=serviceaccounts;services;configmaps,verbs=get;list;watch;create;update;patch;delete
+
+// RBAC for managing Console CRs
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=consoles,verbs=get;patch;list;watch
+// +kubebuilder:rbac:groups=console.openshift.io,resources=consoleplugins,verbs=get;list;watch;create;update;delete;patch
+
+// RBAC for reading cluster version
+// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
+
+func RegisterWithManager(mgr ctrl.Manager, opts Options) error {
+	logger := ctrl.Log.WithName("observability-ui")
+
+	clusterVersion, err := getClusterVersion(mgr.GetAPIReader())
+
+	if err != nil {
+		logger.Error(err, "failed to get cluster version")
+	}
+
+	rm := &resourceManager{
+		k8sClient:      mgr.GetClient(),
+		scheme:         mgr.GetScheme(),
+		logger:         logger,
+		pluginConf:     opts.PluginsConf,
+		clusterVersion: clusterVersion.Status.Desired.Version,
+	}
+
+	generationChanged := builder.WithPredicates(predicate.GenerationChangedPredicate{})
+
+	ctrl, err := ctrl.NewControllerManagedBy(mgr).
+		For(&obsui.ObservabilityUIPlugin{}).
+		Owns(&appsv1.Deployment{}, generationChanged).
+		Owns(&v1.Service{}, generationChanged).
+		Owns(&v1.ServiceAccount{}, generationChanged).
+		Owns(&rbacv1.ClusterRole{}, generationChanged).
+		Owns(&rbacv1.ClusterRoleBinding{}, generationChanged).
+		Owns(&osv1alpha1.ConsolePlugin{}, generationChanged).
+		Build(rm)
+	if err != nil {
+		return err
+	}
+	rm.controller = ctrl
+
+	return nil
+}
+
+func getClusterVersion(k8client client.Reader) (*configv1.ClusterVersion, error) {
+	clusterVersion := &configv1.ClusterVersion{}
+	key := client.ObjectKey{Name: "version"}
+	if err := k8client.Get(context.TODO(), key, clusterVersion); err != nil {
+		return nil, err
+	}
+	return clusterVersion, nil
+}
+
+func consolePluginCapabilityEnabled(k8client client.Reader, name types.NamespacedName) bool {
+	current := &osv1alpha1.ConsolePlugin{}
+	err := k8client.Get(context.TODO(), name, current)
+
+	return err == nil || !metaerrors.IsNoMatchError(err)
+}
+
+func (rm resourceManager) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := rm.logger.WithValues("plugin", req.NamespacedName)
+
+	if !consolePluginCapabilityEnabled(rm.k8sClient, req.NamespacedName) {
+		logger.Info("Cluster console plugin not supported. Skipping observability UI plugin reconciliation")
```

Comment:
IIUC we would enter this branch in one of the following cases:
* COO is deployed on vanilla Kubernetes and the openshift feature gate is enabled. But we should already have failed the reconciler init since it tries to get the cluster version?
* COO is deployed on OpenShift but it has no permissions to get the ConsolePlugin objects.
* COO is deployed on OpenShift but the ConsolePlugin CRD doesn't exist (is it even possible? old version of OCP?).

I wonder if the check shouldn't be done only once at startup? Is there a case where it would return false then true?

Reply:
- I adjusted the code to exit early when the feature gate is enabled but running on vainilla k8s, so it effectively fails the reconciler
- Yes, this could be the case. We can adjust the log message
- Yes, versions before 4.11 don't have the plugin CRD

If we have a cluster upgrade then it can start on false and become true for a supported version.

Reply:
> If we have a cluster upgrade then it can start on false and become true for a supported version.

That currently requires a operator restart though doesn't it? I suppose this scenario is covered in https://issues.redhat.com/browse/COO-106?

@simonpasquier are you ok with this for now or does this need any changes still?